packages/portalnetwork/src/wire/utp/Socket/UtpSocket.ts

```diff
@@ -145,52 +151,69 @@ export class UtpSocket extends EventEmitter {
       return await sendAckPacket(this)
     } else {
       this.state = ConnectionState.Connected
-      if (packet.header.ackNr > 1 && !this.ackNrs.includes(packet.header.ackNr)) {
-        this.ackNrs.push(packet.header.ackNr)
+      if (this.writer) {
+        const inFlight = this.outBuffer.size
+        this.cur_window = inFlight * DEFAULT_PACKET_SIZE
+        const needed = this.dataNrs.filter((n) => !this.ackNrs.includes(n))
+        this.logger(`cur_window: ${this.cur_window} bytes in flight`)
+        this.logger(
+          `AckNr's received (${this.ackNrs.length}/${
+            this.writer.sentChunks.length
+          }): ${this.ackNrs[0]?.toString()}...${
+            this.ackNrs.slice(1).length > 3
+              ? this.ackNrs.slice(this.ackNrs.length - 3)?.toString()
+              : this.ackNrs.slice(1)?.toString()
+          }`
+        )
+        this.logger(`AckNr's needed (${needed.length}/${
+          Object.keys(this.writer.dataChunks).length
+        }): ${needed.slice(0, 3)?.toString()}${
+          needed.slice(3)?.length > 0 ? '...' + needed[needed.length - 1] : ''
+        }
+          `)
       }
-      this.logger(
-        `AckNr's needed: ${this.dataNrs.toString()} \n AckNr's received: ${this.ackNrs.toString()}`
-      )
       if (this.compare()) {
         this.logger(`all data packets acked`)
         return true
       } else {
-        this.logger(`Still waiting for ${this.dataNrs.length - this.ackNrs.length} STATE packets.`)
+        this.writer?.writing && this.writer?.start()
       }
     }
   }
 
+  generateSelectiveAckBitMask(): Uint8Array {
+    const window = new Array(32).fill(false)
+    for (let i = 0; i < 32; i++) {
+      if (this.ackNrs.includes(this.ackNr + 1 + i)) {
+        window[bitmap[i] - 1] = true
+      }
+    }
+    const bitMask = new BitVectorType(32).serialize(BitArray.fromBoolArray(window))
+    return bitMask
+  }
+
   async handleDataPacket(packet: Packet): Promise<Packet> {
     this.state = ConnectionState.Connected
-    if (this.nextSeq !== packet.header.seqNr || this.nextAck !== packet.header.ackNr) {
-      this.logger(
-        `expecting ${this.nextSeq}-${this.nextAck}.  got ${packet.header.seqNr}-${packet.header.ackNr}`
-      )
-    }
-    const expected = this.nextSeq === packet.header.seqNr
-    this.nextSeq = packet.header.seqNr + 1
+    const expected = this.ackNr + 1 === packet.header.seqNr
     this.seqNr = this.seqNr + 1
-    this.ackNr = packet.header.seqNr
     if (!this.reader) {
       this.reader = new ContentReader(this, packet.header.seqNr)
     }
+    this.ackNrs.push(packet.header.seqNr)
     await this.reader.addPacket(packet)
     if (expected) {
-      this.ackNrs.push(this.ackNr)
+      this.ackNr =
+        this.ackNrs.sort((a, b) => a - b).find((n, i) => this.ackNrs[i + 1] !== n + 1) ??
```

Comment:
This feels like a lot of work to find the right ack number if we're having to sort the array every single time.  Is there any difference in result between the two parts of this expression?

Reply:
OK help me think through this...

* `this.ackNr` only changes when the in-order sequence grows.
* `this.ackNrs` does update even with out of order packets, so `this.ackNrs` can end up unsorted.
* An "expected" packet means the next in-order packet is being acked, whether it is at the tip of the sequence or in the middle.  Then this.ackNr changes to the last of the in-order ackNrs, which could still be in the middle of the array of ackNrs.

packets are coming in like:
5...6...7...9...11...
`this.ackNrs = [5, 6, 7, 9, 11]`

9 and 11 were out of order, so got selective-acks, this.ackNr doesn't change.
`this.ackNr = 7`

then we get packet 8, and respond with a regular ACK, with an ackNr that references the new tip of the in-order sequence.
so `this.ackNrs = [5, 6, 7, 9, 11, 8]`
and `this.ackNr` needs to change to *9*, which is the highest of the in-order seqNrs, but it's not the largest number, or the last in the array.
```
        this.ackNrs.sort((a, b) => a - b).find((n, i) => this.ackNrs[i + 1] !== n + 1) ??
```
 
any other ideas on how to find the new ackNr?

Reply:
I'm not sure if this is better or not but what if we change `this.ackNrs` to a `Set`?  Then, the above would change to something like:
```js
let i = this.ackNr + 1
let foundNextAck = false
while (!foundNextAck) {
  if (this.ackNrs.has(i)) {
    i++
  } else {
     foundNextAck = true
     this.ackNr = i
  }
}
```
Then, we're only iterating over the ack values greater than `this.ackNr` and not having to do a bunch of sorting first.  I'm not 100% sure this is any more efficient but it makes sense in my head that it would be faster.

Reply:
Shouldn't need to.
```
const list = [1, 2, 3]
// list.length === 3
list[4] = 5
// list === [1,  2, 3, undefined, 5]
// list.length === 5
// list[3] === undefined
```

Reply:
no pushing involved.